In [ ]:
import os
os.environ['PYKX_4_1_ENABLED'] = "True"
import pykx as kx
kx.config
kx.util.jupyter_qfirst_enable()

\--- Day 2: Gift Shop --- [link](https://adventofcode.com/2025/day/2)
-------------------------

You get inside and take the elevator to its only other stop: the gift shop. "Thank you for visiting the North Pole!" gleefully exclaims a nearby sign. You aren't sure who is even allowed to visit the North Pole, but you know you can access the lobby through here, and from there you can access the rest of the North Pole base.

As you make your way through the surprisingly extensive selection, one of the clerks recognizes you and asks for your help.

As it turns out, one of the younger Elves was playing on a gift shop computer and managed to add a whole bunch of invalid product IDs to their gift shop database! Surely, it would be no trouble for you to identify the invalid product IDs for them, right?

They've even checked most of the product ID ranges already; they only have a few product ID ranges (your puzzle input) that you'll need to check. For example:

`11-22,95-115,998-1012,1188511880-1188511890,222220-222224,  1698522-1698528,446443-446449,38593856-38593862,565653-565659,  824824821-824824827,2121212118-2121212124`

(The ID ranges are wrapped here for legibility; in your input, they appear on a single long line.)

The ranges are separated by commas (,); each range gives its first ID and last ID separated by a dash (-).

Since the young Elf was just doing silly patterns, you can find the invalid IDs by looking for any ID which is made only of some sequence of digits repeated twice. So, 55 (5 twice), 6464 (64 twice), and 123123 (123 twice) would all be invalid IDs.

None of the numbers have leading zeroes; 0101 isn't an ID at all. (101 is a valid ID that you would ignore.)

Your job is to find all of the invalid IDs that appear in the given ranges. In the above example:

*   11-22 has two invalid IDs, 11 and 22.
    
*   95-115 has one invalid ID, 99.
    
*   998-1012 has one invalid ID, 1010.
    
*   1188511880-1188511890 has one invalid ID, 1188511885.
    
*   222220-222224 has one invalid ID, 222222.
    
*   1698522-1698528 contains no invalid IDs.
    
*   446443-446449 has one invalid ID, 446446.
    
*   38593856-38593862 has one invalid ID, 38593859.
    
*   The rest of the ranges contain no invalid IDs.
    

Adding up all the invalid IDs in this example produces 1227775554.

What do you get if you add up all of the invalid IDs?

In [165]:
show line:first read0`:test.txt

"11-22,95-115,998-1012,1188511880-1188511890,222220-222224,1698522-1698528,44..


In [166]:
"," vs line

"11-22"
"95-115"
"998-1012"
"1188511880-1188511890"
"222220-222224"
"1698522-1698528"
"446443-446449"
"38593856-38593862"
"565653-565659"
"824824821-824824827"
"2121212118-2121212124"


In [167]:
show input:{"J"$"-" vs x} each "," vs line

11         22        
95         115       
998        1012      
1188511880 1188511890
222220     222224    
1698522    1698528   
446443     446449    
38593856   38593862  
565653     565659    
824824821  824824827 
2121212118 2121212124


In [168]:
/ generate all numbers in range - use pattern matching form 4.1 for params
{[(mn;mx)] mn+til 1+mx-mn} 0N!input 0

11 12 13 14 15 16 17 18 19 20 21 2211 22



# Part 1
Rely on even-digitness and divisibility of 12341234 by 10001

In [169]:
/ only even-number-digit numbers are of interest
{[(mn;mx)] s where 0=mod[ceiling 10 xlog s:mn+til 1+mx-mn;2]} 0N!input 2

1001 1002 1003 1004 1005 1006 1007 1008 1009 1010 1011 1012
998 1012


In [170]:
/ symmetrical number abab (3rd power of 10) should be divisible by 101: 1+10^(1+3)/2
{[(mn;mx)] s where (0=mod[s;1+10 xexp n2 div 2])&0=mod[n2:ceiling 10 xlog s:mn+til 1+mx-mn;2]} input 2

,1010


In [171]:
/ solution
sum raze {[(mn;mx)] s where (0=mod[s;1+10 xexp n2 div 2])&0=mod[n2:ceiling 10 xlog s:mn+til 1+mx-mn;2]} each input

1227775554


\--- Part Two ---
-----------------

The clerk quickly discovers that there are still invalid IDs in the ranges in your list. Maybe the young Elf was doing other silly patterns as well?

Now, an ID is invalid if it is made only of some sequence of digits repeated at least twice. So, 12341234 (1234 two times), 123123123 (123 three times), 1212121212 (12 five times), and 1111111 (1 seven times) are all invalid IDs.

From the same example as before:

*   11-22 still has two invalid IDs, 11 and 22.
    
*   95-115 now has two invalid IDs, 99 and 111.
    
*   998-1012 now has two invalid IDs, 999 and 1010.
    
*   1188511880-1188511890 still has one invalid ID, 1188511885.
    
*   222220-222224 still has one invalid ID, 222222.
    
*   1698522-1698528 still contains no invalid IDs.
    
*   446443-446449 still has one invalid ID, 446446.
    
*   38593856-38593862 still has one invalid ID, 38593859.
    
*   565653-565659 now has one invalid ID, 565656.
    
*   824824821-824824827 now has one invalid ID, 824824824.
    
*   2121212118-2121212124 now has one invalid ID, 2121212121.
    

Adding up all the invalid IDs in this example produces 4174379265.

What do you get if you add up all of the invalid IDs using these new rules?

2 digits - only 11  -> aa

3 digits - only 111 -> aaa

4 digits - 101 and 1111 -> abab and aaaa

5 digits - only 11111 -> aaaaa

6 digits - 1001, 10101 and 111111 -> abcabc, ababab and aaaaaa

*in general it is pair-wise factorisations*

In [172]:
/example for 12
111111111111 div 1 / 12 1-digits
121212121212 div 12 / 6 2-digits
123123123123 div 123 / 4 3-digits
123412341234 div 1234 / 3 4-digits
123456123456 div 123456 / 2 6-digits

111111111111
10101010101
1001001001
100010001
1000001


In [173]:
/ simple factorisation
ffn:{-1_flip[(x div d;d)]idx:where 0=x mod/:d:1+til x}
show factors:ffn 12

12 1
6  2
4  3
3  4
2  6


In [174]:
/ generate "Nth power of 10 shited m times" pattern
{[d;p] 1+{y+x*y}/[p10;(d-2)#p10:prd p#10]}[4;3]

1001001001


In [175]:
/ example for 12, 10, 3 and 2
{[d;p] 1+{y+x*y}/[p10;(d-2)#p10:prd p#10]}.'ffn 12
{[d;p] 1+{y+x*y}/[p10;(d-2)#p10:prd p#10]}.'ffn 10
{[d;p] 1+{y+x*y}/[p10;(d-2)#p10:prd p#10]}.'ffn 3
{[d;p] 1+{y+x*y}/[p10;(d-2)#p10:prd p#10]}.'ffn 2

111111111111 10101010101 1001001001 100010001 1000001
1111111111 101010101 100001
,111
,11


In [176]:
/ pre-generate divisors
k:1_1+til max ceiling 10 xlog input[;1]
d:k!{{[d;p] 1+{y+x*y}/[p10;(d-2)#p10:prd p#10]}.'ffn x}each k
d

2 | ,11
3 | ,111
4 | 1111 101
5 | ,11111
6 | 111111 10101 1001
7 | ,1111111
8 | 11111111 1010101 10001
9 | 111111111 1001001
10| 1111111111 101010101 100001


In [177]:
/ getting divisors 
{[(mn;mx)] d ceiling 10 xlog s:0N!mn+til 1+mx-mn} input 2

,111
,111
,111
1111 101
1111 101
1111 101
1111 101
1111 101
1111 101
1111 101
1111 101
1111 101
1111 101
1111 101
1111 101
998 999 1000 1001 1002 1003 1004 1005 1006 1007 1008 1009 1010 1011 1012


In [178]:
/ any divisor getting 0 remainder?
raze {[(mn;mx)] s where any each 0=s mod'd ceiling 10 xlog s:mn+til 1+mx-mn} input 2

999 1010


In [179]:
/ solution 2
sum raze {[(mn;mx)] s where any each 0=s mod'd ceiling 10 xlog s:mn+til 1+mx-mn} each input

4174379265
